### ISSUE TITLE
Proposal: Engineering a Common Plugin Infrastructure

### ISSUE THREAD

**Issue by deliciouslytyped** [NixOS/nixpkgs#59344]
There are various separate, probably incompatible implementations of plugin infrastructure across nixpkgs, examples being at least:

- TODO, list here

The more advanced implementations usually (?) use some kind of fixpoint mechanism which can be quite hard to understand.

Would it be possible and reasonable to factor out a plugin handling API that can be specialized to various desired implementations(, such that the internal mechanisms do not need to be understood by the user)? This could also result in the overall simplification of these components of nixpkgs.

A generic plugin system should:
- provide one or more (possibly extensible) methods of applying plugins to a package:
  - feeding plugins back into the original derivation, e.g.
    `{..., plugins ? ... }: mkDerivation ...`
  - merging base and plugin derivations via something like `symlinkJoin`.
    This has the advantage of being faster and less wasteful than rebuilding the base package, but may fail with improper symlink handling?
- provide the ability to have in-tree (nixpkgs) plugins
- provide the ability to have out-of-tree plugins via overrides, overlays, or some similar mechanism
- ...other...?

**Comment by deliciouslytyped** [comment]
If a desirable solution already exists, then other plugin systems listed in the (TODO) aggregation should be migrated to it.

**Comment by deliciouslytyped** [comment]
This post: syntax/semantics/ergonomics

I quite like the `rootPackage.withPackages (p: with p; [ a b c ])` pattern.

Under what circumstances should it be considered meaningful to include non-child packages of the rootPackage in the list? Are there conditions where it's useful to pass in external packages? Say, something like `python37.withPackages (p: [ pkgs.hello ])`, or should the intended result of such a thing be achieved some other way?

This should not be the same as accepting out-of-tree plugins; I think it would be reasonable to accept something like functions in the plugin list, where during evaluation the root package is applied to the plugins if necessary, and passed on to another builder (or some similar approach).
This is done in my current implementation of ghidra plugins because they need access to a build script located in the root package. (TODO substantiating example)

Alternatively, external overrides should be applied strictly through some pattern like the following:
`(rootPackage.overridePackages { d = ...; }).withPackages (p: with p; [ a b c d ])`
Though I find the parenthesization a bit awkward.

**Comment by deliciouslytyped** [comment]
Per suggestions I have attempted to explore `makeScope`/`newScope`/... , but I haven't been able to understand them yet enough to get them to work right. My current understanding is that would at least cover my minimum desired functionality for providing overrides.

**Comment by joepie91** [comment]
Some quick additional notes before I forget, though I don't have much time to elaborate right now:

- Mods for games are also plugins, not sure whether those require special handling.
- There's currently a disconnect in cases where plugins can be managed from within the application as well as from Nix configuration; there should probably be some sort of way to reconcile that, though it may involve extra tooling.
- PlatformIO orchestrates a bunch of different tools for interacting with different embedded devices and architectures; it currently downloads those on runtime. Not sure if that would fall under the conceptual meaning of 'plugins' here.

**Comment by eighal** [comment]
Packages managed by the package managers of (at least interpreted) languages, like Ruby gems, are also plugins, right?

**Comment by joepie91** [comment]
More: Jekyll has plugins.

**Comment by deliciouslytyped** [comment]
@joepie91 
To address your points in some order:

2)
The application managed plugins problem is something that has bothered me for a while and I'm not sure what to do about it. 

My knee-jerk reaction is that applications need to support application-external plugin/setting management if we don't want strange things to happen? The reason I mention settings management, is because that's usually how plugins get added to the "current set of running plugins", right? My mental model is currently based off of firefox - for which I'm not sure of the details (but there's this mess of notes https://github.com/rycee/home-manager/issues/606 ) - which is to say, I think (?) I'm agreeing with you.

1) I don't see yet why game mods are special? I'm imagining that specific installation functionality would be handled by domain specific handlers anyway. Application specific hooks.

3) I will talk to you about this on IRC and update this post.

@eighal yeah I'm inclined to think so. My current heuristic for "plugins" is whether there is some central thing that needs to have plugins fed back into it, that being the interpreter/'s package environment.
But maybe I'm going about this from the wrong direction. Instead of thinking about plugins, maybe we should be thinking about "things that have a domain specific set of 'sub'packages, that must be fed-back the packages"? I wouldn't be surprised if this basically reduces to the thing going on with the fixpoint stuff.

Misc:
added ruby and jekyll to tentative list.

**Comment by deliciouslytyped** [comment]
TL;DR: if it has something that looks like a `.withPackages (p: ...)`, or looks like it would wan't a `.withPackages (p: ...)` then at least that is relevant.

**Comment by aanderse** [comment]
The `redmine` service has plugins and themes. This is currently not implemented well.

**Comment by coreyoconnor** [comment]
Projects with internal package managers for handling plugins I've worked with: `jenkins`, `retroarch`, `gnome-shell`. These all can download their own plugins/extensions which, sometimes, works. Other times there are requirements to the .so/.jar/.js that are not part of the provided environment. Hence: those need to be built via nix and, preferably, block the installation of built artifacts dynamically.

As @deliciouslytyped mentions: applications themselves need some consideration to prevent bad things happening. 

EG: in `retroarch` an update for bundled core can be requested but will fail with an odd error as that .so is RO.

**Comment by coreyoconnor** [comment]
I noticed the question mark next to jenkins and decided to dig in a bit mroe. 

* The relevant code is at nixos/modules/services/continuous-integration/jenkins/default.nix 
* the tool referenced `jenkinsPlugins2nix` does not (afaict) *build* the plugin. Only packages a pre-built plugin.
* only runtime dependencies would impact actually using the plugin
* jenkins itself is not built by nix but only packaged

I agree this puts jenkins in a different group than the above.

**Comment by joepie91** [comment]
Addition: KDE has a plugin manager as well, for "plasma widgets" and such, I believe.

**Comment by deliciouslytyped** [comment]
@coreyoconnor The reason I "separated" it was because on a cursory look I didn't see an implementation of handling plugins - hence: "No implementation". If it had an implementation I would have listed it under the rest. (This is just a comment on that and not really adressing anything else.)

**Comment by deliciouslytyped** [comment]
I'm currently poking around overlays and...that stuff. Still trying to figure out how it works.

**Comment by deliciouslytyped** [comment]
I have a preliminary system for Ghidra I will try upstreaming soon, otherwise I'm still exploring the problem space.

**Comment by deliciouslytyped** [comment]
https://github.com/NixOS/nixpkgs/issues/45269

**Comment by deliciouslytyped** [comment]
Possibly related https://github.com/NixOS/rfcs/pull/3

**Comment by aanderse** [comment]
Just noticed another item which might be of interest. `squirrel-sql` has database drivers that can be considered plugins. See `nixpkgs/pkgs/development/tools/database/squirrel-sql/default.nix` for details.

**Comment by deliciouslytyped** [comment]
Misc wishlist: It would be nice if `enableDebugging` worked properly across package sets (where appropriate).

**Comment by deliciouslytyped** [comment]
Forgot to link my WIP POC: https://github.com/NixOS/nixpkgs/pull/60664/

**Comment by deliciouslytyped** [comment]
I turned it into a WIP library https://github.com/deliciouslytyped/nix-rootedoverlay/ , it's pretty broken right now. Help wanted.

**Comment by aanderse** [comment]
As a new data point I found that `moodle` plugins are scattered all throughout the directory structure of the package. Typically plugins for an application should go under 1 or maybe 2 different directories (say `plugins/` and/or `themes/`, for example), but with `moodle` they go all over the place, depending on what type of plugin you're looking at.

For example plugins classified as `Enrollment plugins` go under `enrol/`, while those classified as `Blocks` go under `blocks/`, etc... At this time it appears `moodle` supports upwards to 50 different types of plugins, each with a different folder to be installed under.

**Comment by deliciouslytyped** [comment]
Where can I find the moodle stuff?

From the perspective of my WIP rootedoverlay stuff:
Not sure which of the two you're trying to point out here? - I'm assuming the latter:
- From a packaging standpoint personally I'd just shove everything in a top level plugins directory and the top level plugin list can handle exposing things to the user
- From the standpoint of "applying" plugins, I this would still fit into a supplying a custom `Root -> Plugin -> Root` function, that can handle individual plugins specifying how they should be installed.

Does that sound reasonable?

**Comment by aanderse** [comment]
@deliciouslytyped I was mostly just providing another data point... but yes, the second point sounds good :smile:

**Comment by deliciouslytyped** [comment]
All good, thanks :D

**Comment by deliciouslytyped** [comment]
I'm currently working on a POC for the R package set.

**Comment by deliciouslytyped** [comment]
There's a somewhat broken implementation at https://github.com/deliciouslytyped/nix-rer/ currently.
The major bugs off the top of my head are
- propagated inputs aren't handled correctly (I'm not sure how the original implementation dealt with this) so `surrogate.cran.curl` doesn't build if it's the dependency of something, but it builds by itself. 
- some packages are not found both because I don't have a way to search the global namespace yet, and I haven't added the bioc package sets.
Also the codebase is still quite a mess.

**Comment by deliciouslytyped** [comment]
https://discourse.nixos.org/t/wip-rfc-a-new-nixpkgs-frontend-for-language-infrastructures/3447

**Comment by deliciouslytyped** [comment]
@barakber I'm thinking of writing a library set for Prolog. I'm not yet familiar with how prolog libraries work, can you maybe explain a bit how https://github.com/NixOS/nixpkgs/blob/9239460aa11071ee3d1d22038e71e818087739f7/pkgs/development/compilers/swi-prolog/default.nix#L11 is supposed to work? Thanks!

**Comment by deliciouslytyped** [comment]
I'm currently working on porting the Idris infrastructure.

**Comment by barakber** [comment]
Sure.  `packInstall` is a Nix function that is using https://www.swi-prolog.org/pldoc/man?section=prologpack from the shell, to install a list of "packs" (from https://www.swi-prolog.org/pack/list) during the installation phase of the SWI-prolog Nix derivation.
The goal is really to have a "swipl.withPackages" that does all pack_install()s for you.

**Comment by deliciouslytyped** [comment]
https://github.com/NixOS/nixpkgs/pull/46842

**Comment by deliciouslytyped** [comment]
It seems like a good idea to read through and steal some ideas from https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md , there appear to have been developments since I last read it.

**Comment by deliciouslytyped** [comment]
TODO: read through https://discourse.nixos.org/t/scaling-the-python-package-set-in-nixpkgs/3749/ , evaluate relevance in both directions, steal ideas

**Comment by deliciouslytyped** [comment]
TODO: look at whats going on in https://github.com/NixOS/nixpkgs/issues/53834#issue-398544390

**Comment by toraritte** [comment]
This discussion seems very relevant: [Status of lang2nix approaches](https://discourse.nixos.org/t/status-of-lang2nix-approaches/14477)

**Comment by deliciouslytyped** [comment]
If that's relevant, the dream2nix project may be as well.
